libsweep/src/sweep.cc

```diff
@@ -46,149 +41,102 @@ void sweep_error_destruct(sweep_error_s error) {
   delete error;
 }
 
-sweep_device_s sweep_device_construct_simple(sweep_error_s* error) {
-  SWEEP_ASSERT(error);
-
-  return sweep_device_construct("/dev/ttyUSB0", 115200, error);
-}
-
-sweep_device_s sweep_device_construct(const char* port, int32_t bitrate, sweep_error_s* error) {
+// internal implementations of functions - can throw exceptions
+namespace {
+
+struct Error : ::sweep::ErrorBase {
+  using ::sweep::ErrorBase::ErrorBase;
+};
+
+// forward declarations for internal definitions
+sweep_device_s sweep_device_construct_simple();
+sweep_device_s sweep_device_construct(const char* port, int32_t bitrate);
+void sweep_device_destruct_impl(sweep_device_s device);
+void sweep_device_start_scanning(sweep_device_s device);
+void sweep_device_stop_scanning(sweep_device_s device);
+sweep_scan_s sweep_device_get_scan(sweep_device_s device);
+int32_t sweep_device_get_motor_speed(sweep_device_s device);
+void sweep_device_set_motor_speed(sweep_device_s device, int32_t hz);
+int32_t sweep_device_get_sample_rate(sweep_device_s device);
+void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz);
+void sweep_device_reset(sweep_device_s device);
+
+sweep_device_s sweep_device_construct_simple() { return sweep_device_construct("/dev/ttyUSB0", 115200); }
+
+sweep_device_s sweep_device_construct(const char* port, int32_t bitrate) {
   SWEEP_ASSERT(port);
   SWEEP_ASSERT(bitrate > 0);
-  SWEEP_ASSERT(error);
-
-  sweep::serial::error_s serialerror = nullptr;
-  sweep::serial::device_s serial = sweep::serial::device_construct(port, bitrate, &serialerror);
-
-  if (serialerror) {
-    *error = sweep_error_construct(sweep::serial::error_message(serialerror));
-    sweep::serial::error_destruct(serialerror);
-    return nullptr;
-  }
 
-  auto out = new sweep_device{serial, /*is_scanning=*/true};
+  sweep::serial::device_s serial = sweep::serial::device_construct(port, bitrate);
 
-  sweep_error_s stoperror = nullptr;
-  sweep_device_stop_scanning(out, &stoperror);
+  auto out = std::unique_ptr<sweep_device>(new sweep_device{serial, /*is_scanning=*/true});
 
-  if (stoperror) {
-    *error = stoperror;
-    sweep_device_destruct(out);
-    return nullptr;
-  }
+  sweep_device_stop_scanning(out.get());
 
-  return out;
+  return out.release();
 }
 
-void sweep_device_destruct(sweep_device_s device) {
+void sweep_device_destruct_impl(sweep_device_s device) {
```

Comment:
To avoid naming collision (the API I version doesn't hav an error parameter to distinguish between the two). See comment below about why I have two in the first place